README.md
# Vaadin Medium Editor [![Bintray](https://img.shields.io/bintray/v/moberwasserlechner/maven/vaadin-medium-editor.svg)](https://bintray.com/moberwasserlechner/maven/vaadin-medium-editor/_latestVersion) [![PayPal](https://img.shields.io/badge/%24-donate-0CB3EB.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=N8VS2P9233NJQ) [![License](https://img.shields.io/badge/license-MIT-B34ED4.svg)](https://github.com/moberwasserlechner/vaadin-medium-editor/blob/master/LICENSE)

Vaadin 8 wrapper for the medium editor javascript library. https://github.com/yabwe/medium-editor

Important: For Vaadin 7 please use [1.2.2](https://github.com/moberwasserlechner/vaadin-medium-editor/releases/tag/1.2.2)

## Features

* MediumEditor component
* MediumEditorField for simple usage in `com.vaadin.data.fieldgroup.BeanFieldGroup<T>` or `com.vaadin.data.fieldgroup.FieldGroup` 
* Configure the editor's options with a fluent api
* Usage of Vaadin's FontAwesome integration, therefore no addional css file must be included.
* Localization for English and German 
* Custom localization by configuration 

## Demo

### Vaadin Addon

* http://vaadin-demos.qqjtxeeuih.eu-central-1.elasticbeanstalk.com:5700

If you want to run the demo application locally, see the [Contribution Section](#run-the-demo-local)

### Medium Editor

* http://yabwe.github.io/medium-editor/
* https://yabwe.github.io/medium-editor/demo.html

## Installation

### Download

[![Bintray](https://img.shields.io/bintray/v/moberwasserlechner/maven/vaadin-medium-editor.svg)](https://bintray.com/moberwasserlechner/maven/vaadin-medium-editor/_latestVersion)

### Vaadin Directory

Get the addon from 
https://vaadin.com/directory#!addon/medium-editor-wrapper. 

Vaadin runs its own Maven repository and you can download the addon there as well, but you will need to create a free vaadin account first.

### Maven

Repository

    <repositories>
      <repository>
        <snapshots>
          <enabled>false</enabled>
        </snapshots>
        <id>central</id>
        <name>bintray</name>
        <url>http://jcenter.bintray.com</url>
      </repository>
    </repositories>
    
Dependency

    <dependencies>
      <dependency>
        <groupId>com.byteowls</groupId>
        <artifactId>vaadin-medium-editor</artifactId>
        <version>replace.with.version</version>
      </dependency>
    </dependencies>


### Gradle

Repository

    repositories {
      jcenter()
    }
     
Dependency

    dependencies {
      compile ("com.byteowls:vaadin-medium-editor:replace.with.version")
    }
## Usage

For more examples please see the [demo app](#vaadin-addon)

### MediumEditor
  
    MediumEditor editor = new MediumEditor();
    editor.setFocusOutlineEnabled(false);
    editor.setJsLoggingEnabled(true);
    // using lorem ipsum to get some text. see demo for dependency
    editor.setContent(Lorem.getHtmlParagraphs(3, 3));
    editor.addBlurListener(value -> {
      preview.setValue(value);
    });
    
The MediumEditor is a vaadin component in its most basic form a `<div>`. So in contrast to the javascript library under the hood, there is no constructor for setting a selector string or a inserting dom elements.

This is a known limitation because you cannot have a single editor configuration for more than one field.
    
    // inherited vaadin component method
    editor.setSizeFull();
    
Other wrapper specific options.
    
* focusOutlineEnabled ... If true the outline / border is shown as soon as the components gets the focus.
* jsLoggingEnabled ... If true logging with `console.log` in the connector script is enabled. Setting this to true, let you review
* content
* addBlurListener ... On every blur event on client side the current content is sent to the server. Register your listener to do sth with it.   

You're able to use the MediumEditor just like that

    protected void init(VaadinRequest request) {
      MediumEditor editor = new MediumEditor();
      setContent(editor);
    }

In this case the default medium-editor options are set. See https://github.com/yabwe/medium-editor#mediumeditor-options for the javascript lib options.

You can overwrite the default options with 

    MediumEditor editor = new MediumEditor();
    editor.configure(
        editor.options()
        // use FontAwesome button labels
        .fontawesomeButtonLabels()
        // start configuring the toolbar
        .toolbar()
          // only this buttons should be included
          .buttons(Buttons.BOLD, Buttons.ITALIC, Buttons.H1, Buttons.JUSTIFY_CENTER)
          // the the german translations for the buttons. be aware of the order
          .buttonTranslations("fett", "kursiv", "Ueberschrift1", "zentriert")
          // configuring the button is done, we want to go back and continue configuring other options
          .done()
        // start configuring the placeholder if the component has no content
        .placeholder()
          // the actual text
          .text("Input prompt")
          // configuring the placeholder is done, again we go back and continue configuring other options
          .done()
        // urls added to the content are automatically converted to links 
        .autoLink(true)
        // image dragging
        .imageDragging(false)
        // we're done :)
        .done()
        ); 

The wrappers fluent option api is mostly named after the javascript options documented at https://github.com/yabwe/medium-editor#mediumeditor-options.

I'll probably add more javadoc in the next versions.

### MediumEditorField

The MediumEditorField is great for usage within FieldGroups.

    MediumEditorField editorField = new MediumEditorField(this.i18n.get("messaging.email.content"));
    // get the field's editor for further configuration
    MediumEditor editor = editorField.getEditor();

The `editor` field is then bound to a FieldGroup and might be configured the same way seen above.


## Prerequisite

### Addon
* JDK 8 or higher
* Vaadin 8.0.0 or higher

### Demo
* JDK 8 (because of Lambdas)
* Vaadin 8.0.x


## Missing something?

The Vaadin-Medium-Editor is only a wrapper. So if you have any feature requests or found any bugs in the javascript lib please use Medium Editor's issue tracker https://github.com/yabwe/medium-editor/issues 

In all other cases please create a issue at https://github.com/moberwasserlechner/vaadin-medium-editor/issues or contribute to the project yourself. For contribution see the next section.

## Contribute

### Setup Eclipse

1. Fork repo
2. Open command line
3. Clone your fork `git@github.com:USERNAME/vaadin-medium-editor.git`
4. `cd vaadin-medium-editor`
5. Build eclipse meta data `./gradlew cleanEclipse eclipse`
6. Open Eclipse
7. File -> Import... -> General -> Existing Projects into Workspace
8. Browse to your git repository
9. Check the option "Search for nested projects"
10. Check all 3 projects
11. Press finish

This should take not more than 1-2 minutes. You does not need to use any gradle eclipse plugins. 

### Fix a bug or create a new feature

Please do not mix more than one issue in a feature branch. Each feature/bugfix should have its own branch and its own Pull Request (PR).

1. Create a issue and describe what you want to do at [Issue Tracker](https://github.com/moberwasserlechner/vaadin-medium-editor/issues)
2. Create your feature branch (`git checkout -b feature/my-feature` or `git checkout -b bugfix/my-bugfix`)
3. Test your changes to the best of your ability.
4. Add a demo view to the demo application 
5. Commit your changes (`git commit -m 'Describe feature or bug'`)
6. Push to the branch (`git push origin feature/my-feature`)
7. Create a Github Pull Request

### Run the demo local

The demo application is based on Spring Boot. So its possible to run the Demo as Java Application right out of Eclipse, there is not servlet container needed as Spring Boot has a embedded Tomcat 8 included.

1. Open "Debug Configurations..." dialog
2. Create a new "Java Application"
3. Choose the "vaadin-medium-editor-demo" project
4. Use "com.byteowls.vaadin.mediumeditor.demo.AddonDemoApplication" as Main class
5. Set `-Dprofile=dev` as VM argument. This ensures that source code panel in the demo is correctly filled while developing.
6. Browse to `http://localhost:8080/`

### Translations

Additional languages are very welcome.

ResourceBundles with properties files for the actual translations are used. They are located at

    https://github.com/moberwasserlechner/vaadin-medium-editor/tree/master/addon/src/main/resources/com/byteowls/vaadin/mediumeditor/options
 
Supported languages are:

* English (fallback)
* German
 
### Code Style

Please use the sun coding convention. Please do not use tabs at all!

## License

MIT. Please see [LICENSE](https://github.com/moberwasserlechner/vaadin-medium-editor/blob/master/LICENSE).

## Change Log

Please see [CHANGELOG](https://github.com/moberwasserlechner/vaadin-medium-editor/blob/master/CHANGELOG.md).

CHANGELOG.md
# Changelog

### [2.0.0](https://github.com/moberwasserlechner/vaadin-medium-editor/compare/1.2.2...2.0.0)

* #27 Upgrade to Vaadin 8 and Java 8. For Vaadin 7 use 1.2.2
* #25 Toolbar behind window
* #28 Upgrade to Medium-Editor JS 5.23.0

### [1.2.2](https://github.com/moberwasserlechner/vaadin-medium-editor/compare/1.2.1...1.2.2)

### [1.2.1](https://github.com/moberwasserlechner/vaadin-medium-editor/compare/1.2.0...1.2.1)

* Same as 1.2.0 but invalid manifest for vaadin directory.

### [1.2.0](https://github.com/moberwasserlechner/vaadin-medium-editor/compare/1.1.0...1.2.0)

* #22 Migrate demo application to spring boot, remodel the ui and add more examples
* #23 Server side text change is not assigned to editor if it is in readOnly mode 
* #19 Upgrade to latest medium editor 5.22.0
* #24 Improve contribution section. Instructions for project setup, running demo application locally,... 

### [1.1.0](https://github.com/moberwasserlechner/vaadin-medium-editor/compare/1.0.2...1.1.0)

* #11 Translations for buildin buttons in German and English. Other languages are very welcome.
* #17 Fontawesome buttons labels are enabled by default
* Added no arg constructor to MediumEditorField
* Rename "BuildinButton" enum to "Buttons" because its shorter and its irrelevant if a button is buildin because supported extensions are added the same way. Breaking change.
* #15 Use medium-editor 5.14.4
* #18 Placeholder does not disappear on serverside content change
* #14 Remove Buttons.IMAGE because imHO its not usable
* Change option state from Bean to JsonValue, because options with multiple datatypes and parameter named like java reserved words (like static) are supported now.
* Add FontSize change button
* #1 Support medium editor themes

### [1.0.2](https://github.com/moberwasserlechner/vaadin-medium-editor/compare/1.0.1...1.0.2)

* Fix problem in Safari because of user-select=none on Valo Theme v-app style.
* Demo: Name editor and preview panels

### [1.0.1](https://github.com/moberwasserlechner/vaadin-medium-editor/compare/1.0.0...1.0.1)

* Fix vaadin directory upload restrictions
 * Addon name must not start with "Vaadin"
 * Zip manifest misses 'Vaadin-Package-Version' attribute. see https://github.com/johndevs/gradle-vaadin-plugin/issues/228
* Add Vaadin Directory to installation section in README
* Add Changelog link to README
* Fixing default ANCHOR Button causing an NPE because not implemented
* New allButtons() method for ToolbarBuilder
 


addon/src/main/resources/VAADIN/mediumeditor/js/medium-editor-connector.js
window.com_byteowls_vaadin_mediumeditor_MediumEditor = function() {
	// see the javadoc of com.vaadin.ui.AbstractJavaScriptComponent for all functions on this.
	var THEME_ID = "medium-editor-theme-id";
	var mediumEditor;
	var mediumThemeLink;
	var e = this.getElement();
	// Please note that in JavaScript, this is not necessarily defined inside callback functions and it might therefore be necessary to assign the reference to a separate variable
	var self = this;
	var loggingEnabled = false;
	var focusOutlineEnabled = true;
	var readOnly = false;

	// called every time MediumEditorState is changed
	this.onStateChange = function() {
		var state = this.getState();
		loggingEnabled = state.loggingEnabled;
		focusOutlineEnabled = state.focusOutlineEnabled;

		// #1 theme
		if (typeof mediumThemeLink === 'undefined') {
			// check if a theme css is already loaded
			mediumThemeLink = document.getElementById(THEME_ID);
			if (mediumThemeLink == null) {
				mediumThemeLink = document.createElement("link");
				mediumThemeLink.setAttribute("rel", "stylesheet");
				mediumThemeLink.setAttribute("type", "text/css");
				mediumThemeLink.setAttribute("id", THEME_ID);
				document.getElementsByTagName("head")[0].appendChild(mediumThemeLink);
			}
		}
		var url = this.translateVaadinUri("vaadin://mediumeditor/css/"+state.theme.toLowerCase()+".min.css");
		mediumThemeLink.setAttribute("href", url);

		if (loggingEnabled) {
			console.log("medium-editor: setting value to\n" + state.content);
		}

		if (!focusOutlineEnabled) {
			e.style.outline = "none";
		}

		if (typeof mediumEditor === 'undefined') {
			if (loggingEnabled) {
				console.log("medium-editor: init editor");
			}

			// #12 enable text selection on contenteditable for Safari
			if (loggingEnabled) {
				console.log("medium-editor: enable user-select on contenteditable div because it might be deactivated by valo");
			}
			e.style.webkitUserSelect = "text";
			e.style.userSelect = "text";
			e.style.MozUserSelect = "text";
			e.style.msUserSelect = "text"; // For IE10+

			// ensure that pressing return or enter creates a new line
			e.addEventListener("keydown", function(event) {
				if (event.which == 13 || event.keyCode == 13) {
					event.stopPropagation();
				}
			});

			e.addEventListener("blur", function() {
				if (!readOnly) {
					var val = e.innerHTML;
					if (loggingEnabled) {
						console.log("medium-editor: value on blur is\n" + val);
					}
					self.onValueChange(val);
				}
			});

			if (loggingEnabled) {
				console.log("medium-editor: options are\n", JSON.stringify(state.optionsJson, null, 2));
			}
			mediumEditor = new MediumEditor(e, state.optionsJson);
		}
		var serverContent = state.content || "";
		if (state.readOnly) {
			if (loggingEnabled) {
				console.log("medium-editor: destroying editor because readOnly is enabled!");
			}
			// #23 assign server content to readonly editor.
			e.innerHTML = serverContent;
			mediumEditor.destroy();
		} else {
			// #18 placeholder does not disappear if not using mediumEditor's setContent()
			mediumEditor.setup();
			mediumEditor.setContent(serverContent);
		}
		readOnly = state.readOnly;
	};

}
